### ISSUE TITLE
Spring IDE support for STS 4 plugin for Eclipse

### ISSUE THREAD

**Issue by sanketcs30** [spring-projects/sts4#1163]
Hi @martinlippert , I'm trying to install **STS 3.9.14 plugin in Eclipse IDE** , then it's giving multiple errors and I have resolved these by installing new STS 4 plugin , but the main purpose of using 3.9.14 version STS plugin is to work with spring based projects with XML configurations , there is **need of using spring bean configuration file , namespaces** which is not available in STS 4 plugin , so is Spring team planning for the **same spring IDE support in STS 4** also , it will be really helpful as many **spring based legacy projects** are still going in XML and Java based configurations

Thanks

#STS 4 issue #Support for Spring IDE features @martinlippert

**Comment by martinlippert** [comment]
Thanks @sanketcs30 for reaching out to us with this request. The STS 3.x plugins for Eclipse are no longer maintained and especially due to the continuous evolution of the underlying Eclipse platform, the old versions of the STS 3.x plugins for Eclipse don't work anymore with the latest Eclipse versions.

We are encouraging the transition to the Spring Tools 4 for Eclipse and VSCode since a number of years now and your projects that use Spring XML configurations should still work fine with using Spring Tools 4. The piece that I would like to understand in more depth here is what exactly are you missing in the Spring Tools 4? Can you describe those features in more detail and maybe provide sample projects to illustrate those?

There is basic support for working with Spring XML configuration files in Spring Tools 4, like content-assist for bean refs, class names, etc. So what exactly is not (yet) working, what features are you missing the most?

**Comment by sanketcs30** [comment]
Acutally STS 4 is not having options to create spring bean file , adding namespaces for different spring modules like spring core , aop , mvc , transaction which were available in STS 3 plugin

**Comment by martinlippert** [comment]
(I updated the title if this issue accordingly)

**Comment by sanketcs30** [comment]
Yes Support for cresting new Spring XML configuration files with an easy way to select the Spring XML namespaces based on the libraries on the classpath of the project , it is not available in STS 4 , like in STS 3 plugin when we use following steps
 **File --> new --> other ---> spring ---> spring beans configuration file** and after clicking that configuration file option it will create spring bean configuration file , also in that file itself there will be small tabs like namespaces that is not available in sts 4

Like the image showing namespaces tab
![Spring Modules Namespaces](https://github.com/spring-projects/sts4/assets/150379281/eb02c25c-8122-425e-b9d9-28cd990bbb86)

How can we get these options in STS 4 , so it will be helpful if these features of STS 3 are added in STS 4 plugin

**Comment by nealwei2008** [comment]
> Thanks @sanketcs30 for reaching out to us with this request. The STS 3.x plugins for Eclipse are no longer maintained and especially due to the continuous evolution of the underlying Eclipse platform, the old versions of the STS 3.x plugins for Eclipse don't work anymore with the latest Eclipse versions.
> 
> We are encouraging the transition to the Spring Tools 4 for Eclipse and VSCode since a number of years now and your projects that use Spring XML configurations should still work fine with using Spring Tools 4. The piece that I would like to understand in more depth here is what exactly are you missing in the Spring Tools 4? Can you describe those features in more detail and maybe provide sample projects to illustrate those?
> 
> There is basic support for working with Spring XML configuration files in Spring Tools 4, like content-assist for bean refs, class names, etc. So what exactly is not (yet) working, what features are you missing the most?

Hi @martinlippert , may I ask how to work with Spring XML configuration files in Spring Tools 4? I can't find this option from the pure latest version of STS4 (4-4.21.0.RELEASE) with no plugin.

![Uploading 2023-12-22 17-44-50 的屏幕截图.png…]()

**Comment by sanketcs30** [comment]
Hi @nealwei2008 @martinlippert , yes in STS 4 we get option to create spring starter project , but **option is not available for spring** which was available in STS 3 version , under spring folder in wizards we were able to **create spring bean configuration file** along with **adding namespaces** for different spring modules which was helpful if someone is working with **XML based spring projects**

Below are some of the screenshots which shows how STS 3 plugin was providing options related to spring bean configuration file

**1. Creating spring bean configuration file through wizard**

![Spring folder in Wizard](https://github.com/spring-projects/sts4/assets/150379281/e69e2edb-41c6-4985-ab52-a7fb4e9d8afe)

**2. Once spring bean configuration file is created , it used to give option to add namespaces**

![Namespaces Selection](https://github.com/spring-projects/sts4/assets/150379281/5130684e-0e53-4c33-b909-4849ba883636)

So, these are the **options that are missing in new STS 4** which I think useful if someone is working with XML based spring projects , because each time **manually creating spring bean configuration file and adding namespaces is not a good option**

**Comment by nealwei2008** [comment]
Thanks @sanketcs30 for your answer, I knew the options about STS 3 and I just want to fix the xml error just like the element 'beans' in STS 4 with "_cvc-elt.1.a: Cannot find the declaration of element 'beans'._"
As we have lots of projects having XML config files, we may have to spend some time with them before refactoring.
I have read the wiki of STS3-Migration and find a possible solution about these Spring XML files which would be found under _Language Servers -> Spring Language Servers -> Spring Boot Language Server -> Spring XML Config_. But question is I can't find _Spring Boot Language Server_. 
I also find an option under the path window -> Preferences -> Spring
 -> XML Config Files -> Enable Spring XML Config files support. It seems that there is nothing to be changed when I check it.
What should I do? How does the _Spring XML Config_ work? Or is there any other methods to solve this error I mentioned above? Looking forward to your answer.
![2023-12-23 22-23-36 的屏幕截图](https://github.com/spring-projects/sts4/assets/51744735/6563f1fd-08b7-47be-86ef-863c35d650b6)

**Comment by sanketcs30** [comment]
Thanks @nealwei2008 for your answer , I have gone through your mentioned steps , yes even after applying those steps like window -> Preferences -> Spring -> XML Config Files -> Enable Spring XML Config files support , I am **not able to see any spring folder in the wizards** , so how can we find that is my question , if anyone is having any idea regarding this then please reply , it will be helpful

**Comment by martinlippert** [comment]
Let me provide some clarify around the Spring XML support that we have in the Spring Tools 4... :-)

Spring Tools 4 provides enhanced content-assist for Spring XML config files if you open them in an XML editor in Eclipse or VSCode, for example for Spring references or class names (to name a few). This is what you enable with the mentioned preference at `Preferences -> Spring -> XML Config Files -> Enable Spring XML Config files support`.

(the preference mentioned in https://github.com/spring-projects/sts4/issues/1163#issuecomment-1868308859 is referring to an old structure of the preferences)

There is no wizard to create new Spring XML config files anymore in the Spring Tools 4. You would have to create a simple plain new XML file (in the same way you would create any other new file in your project). That also means that the Spring Tools 4 don't have the wizards mentioned in https://github.com/spring-projects/sts4/issues/1163#issuecomment-1867952875 anymore that allow you to visually select namespaces to be included.

Since we moved the overall tooling to a new architecture, we will most likely not bring back the old wizards for creating Spring XML config files. Instead, we could think about other ways to support you with the tasks that are mentioned here:

- easily creating new Spring XML config files that allows you to easily select namespaces to be included in the XML file (based on what is configured on the classpath of the project)
- adding namespaces to existing Spring XML config files

Most of this would probably happen inside of the editor itself using content-assist and code actions, e.g. when you create a new XML file, it could be pre-filled with some Spring XML config file body blurb and giving you the option to select the namespaces via some content-assist actions directly in the editor. Just an initial idea, but I would like to think in this direction in case we re-introduce support for that in the Spring Tools 4.

Happy to work with you on some initial thoughts and implementations, get feedback, and work on a few incremental steps here.

**Comment by nealwei2008** [comment]
Yes @martinlippert. That would be great if it can be achieved. Compared to building from scratch, there are also a considerable number of historical projects. At the same time, many newcomers learn about previous construction methods such as  XML Config, which is very helpful for them to master new methods.

- easily creating new Spring XML config files
- adding namespaces manually or automatically 
- recognizing automatically for existing files

These are enough for me and I think they should be enough for projects that have not yet been fully upgraded but are only for temporary use.

**Comment by sanketcs30** [comment]
Thanks @martinlippert @nealwei2008 for your support and brief answers. Yes , if STS 4 will include the following things , then it will be **great in case of working with legacy projects with XML configuration files**
    1. Support for creating spring bean configuration file directly through wizard
    2.  Support for adding namespaces for different spring modules once spring bean configuration file is created

**Comment by martinlippert** [comment]
Here is the new editor-centric support for creating Spring XML config files and adding namespaces to it. The list of namespace suggestions is not (yet) based on the exact Spring modules on the classpath of the project, but just a static list of the namespaces from the core Spring framework project.

https://github.com/spring-projects/sts4/assets/107705/1a63f4d2-4e3d-4258-94df-3a41731a1003

Please note that the Spring framework intend to phase out  XML bean definitions and in particular XML configuration namespaces (as mentioned here: https://github.com/spring-projects/spring-framework/issues/24237#issuecomment-1871435297), so we will most likely not add a lot of additional tooling features in this space anymore. But if there is something that you are missing a lot when working with existing Spring XML config files in your existing projects, let us know and we can see on a case-by-case base what is possible to implement for future versions.

**Comment by sanketcs30** [comment]
Hi @martinlippert can you please **mention steps for how to create that spring-config.xml file** , as I'm trying to work with the **STS 4.21.0 version** , there is no option of creating that xml file like previously we used to create like **File -> New -> Other -> Spring configuration file** 
Can u please mention those **steps to create spring config file** and also please mention **how to add static namespaces** as you have mentioned in the previous post , it will be nice if you add any recordings for that
Thanks

**Comment by martinlippert** [comment]
This will come as part of the `4.21.1` release, so stay tuned.

**Comment by sanketcs30** [comment]
Hi @martinlippert I have installed new **STS version that is 4.21.1** as you have mentioned in the previous post , but I'm **not able** to see any **option in the wizard** to create spring bean configuration file and namespaces to add even static namespaces as you have mentioned , can you please mention **the steps for how to get spring config xml file and add namespaces in the new STS version 4.21.1** like the options that were available in the old STS versions

**1. Creation of spring bean configuration file using wizard**

![Spring folder in Wizard](https://github.com/spring-projects/sts4/assets/150379281/bafdf72e-d755-4d73-a84f-70b1f4004f00)

**2. Selection of namespaces**

![Namespaces Selection](https://github.com/spring-projects/sts4/assets/150379281/1cec6ed5-cc4d-41b6-91dc-ab9e4d7ee958)

**Comment by sanketcs30** [comment]
Hi @martinlippert , as you have mentioned in last post I have **installed STS 4.21.1 version** and tried to create spring bean configuration file , but I was **unable to find the option to add namespaces or spring XML config file skeleton option** as you have mentioned in the last post , can u please help me with **how to get spring XML config file skeleton option** or is there any way to add that option 

# Options that are showing in XML file are:-
![Spring XML Support - Copy](https://github.com/spring-projects/sts4/assets/150379281/e90e3394-ca46-488a-96b5-d72762453333)

**Comment by martinlippert** [comment]
Please make sure that you have enabled the Spring XML config file support in the `Settings -> Spring -> XML Config Files`.

**Comment by sanketcs30** [comment]
Thanks @martinlippert , it's working successfully and I'm really thankful for the support that has been provided by you

![Spring cfg file](https://github.com/spring-projects/sts4/assets/150379281/389438ce-0a64-4cfd-822d-f74c6ff4c520)

**Comment by batobolg** [comment]
I also download last version, it dose not work
<img width="1315" alt="image" src="https://github.com/user-attachments/assets/cd3c0ae9-f7db-44b2-a13a-c5b3f8f854f6">

**Comment by martinlippert** [comment]
@batobolg This looks like a different problem (showing an error when content-assist should show up for types), so please open a separate issue for this.

It would also be good if you could open the Error Log view and and paste the details for the error message here, so that we can get a better understanding of that is wrong here.

It might also be useful if you could enable the logging of the language server process (`Preferences -> Language Servers -> Spring Language Servers -> Logging to Console`). You have to restart the IDE for this change to take effect. Once that is done, the log output from the language server process shows up in the Console view, so the content of that might be useful here as well.

Thanks!

**Comment by boolivar** [comment]
@martinlippert I'm really sorry for writing in closed topic but I can't get classpath XSDs to work.
I was testing STS versions from `4.24.0` up to `4.27.0` with fresh new spring starter webmvc project.
Spring XML Config files support is enabled.

I can generate XML config file, but no Content Assist is available for referenced namespace, `spring-context` as an example:

![Screenshot_1](https://github.com/user-attachments/assets/deb886fe-2e16-43d9-8791-79ac4bdbe5c2)

If I put something from `context` namespace I'm getting an error:

![Screenshot_2](https://github.com/user-attachments/assets/0b1009a9-6a04-4807-9c88-276620f573d9)

If I enable external resources downloading Content Assist works but uses resources from website instead of classpath:

![Screenshot_3](https://github.com/user-attachments/assets/b01ba3d2-0eaf-4fe7-82cd-21b0f98ec796)

`tiles-configurer` is a good example of element that doesn't exist in actual schema:

```
org.springframework.beans.factory.xml.XmlBeanDefinitionStoreException: Line 11 in XML document from class path resource [spring-config.xml] is invalid
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:411) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:347) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.loadBeanDefinitions(XmlBeanDefinitionReader.java:319) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:184) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:220) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.support.AbstractBeanDefinitionReader.loadBeanDefinitions(AbstractBeanDefinitionReader.java:191) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.lambda$loadBeanDefinitionsFromImportedResources$0(ConfigurationClassBeanDefinitionReader.java:391) ~[spring-context-6.2.1.jar:6.2.1]
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:986) ~[na:na]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsFromImportedResources(ConfigurationClassBeanDefinitionReader.java:359) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:148) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:121) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:430) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:290) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:349) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:118) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:791) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:609) ~[spring-context-6.2.1.jar:6.2.1]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.4.1.jar:3.4.1]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:752) ~[spring-boot-3.4.1.jar:3.4.1]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439) ~[spring-boot-3.4.1.jar:3.4.1]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:318) ~[spring-boot-3.4.1.jar:3.4.1]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1361) ~[spring-boot-3.4.1.jar:3.4.1]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1350) ~[spring-boot-3.4.1.jar:3.4.1]
	at com.example.demo.SpringStarterDemoApplication.main(SpringStarterDemoApplication.java:12) ~[main/:na]
Caused by: org.xml.sax.SAXParseException: cvc-complex-type.2.4.c: The matching wildcard is strict, but no declaration can be found for element 'mvc:tiles-configurer'.
	at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.error(ErrorHandlerWrapper.java:135) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:396) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:284) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator$XSIErrorReporter.reportError(XMLSchemaValidator.java:512) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.reportSchemaError(XMLSchemaValidator.java:3596) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.handleStartElement(XMLSchemaValidator.java:2144) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.xs.XMLSchemaValidator.emptyElement(XMLSchemaValidator.java:850) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:353) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:114) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:247) ~[na:na]
	at java.xml/com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:342) ~[na:na]
	at org.springframework.beans.factory.xml.DefaultDocumentLoader.loadDocument(DefaultDocumentLoader.java:77) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadDocument(XmlBeanDefinitionReader.java:441) ~[spring-beans-6.2.1.jar:6.2.1]
	at org.springframework.beans.factory.xml.XmlBeanDefinitionReader.doLoadBeanDefinitions(XmlBeanDefinitionReader.java:399) ~[spring-beans-6.2.1.jar:6.2.1]
	... 23 common frames omitted
```

**Comment by martinlippert** [comment]
Hey @boolivar, thanks for reporting. This is most likely related to https://github.com/spring-projects/sts4/issues/581. Let's see if we can find a quick workaround to deal with this until the mentioned issue is resolved.

**Comment by martinlippert** [comment]
I tried to find a workaround for this and here is what I found working in my case. Would be awesome if you could give this a try as well to see if that works on your side as well.

- I enabled the 'download external resources like referenced xsd' in `Preferences -> XML (Wild Web Developer)`
- I added my own catalog file in `Preferences -> XML (Wild Web Developer) -> Catalogs`

This catalog file looks like this (as an example):

```
<?xml version="1.0"?>
<!DOCTYPE catalog PUBLIC "-//OASIS//DTD Entity Resolution XML Catalog V1.0//EN" "http://www.oasis-open.org/committees/entity/release/1.0/catalog.dtd">
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog">
<system systemId="https://www.springframework.org/schema/mvc/spring-mvc.xsd" uri="jar:file:///Users/<username>/.m2/repository/org/springframework/spring-webmvc/6.1.8/spring-webmvc-6.1.8.jar!/org/springframework/web/servlet/config/spring-mvc.xsd"/>
<system systemId="http://www.springframework.org/schema/mvc/spring-mvc.xsd" uri="jar:file:///Users/<username>/.m2/repository/org/springframework/spring-webmvc/6.1.8/spring-webmvc-6.1.8.jar!/org/springframework/web/servlet/config/spring-mvc.xsd"/>
</catalog>
```

The exact path for the URI in there, of course, depends on your local machine. And it points to one specific version of the `spring-webmvc` library and the exact path to the XSD file in that library file. But this should trigger the XML tooling to use this specific XSD to validate the XML files when they refer to this namespace.

It is not y very nice workaround (lots of custom stuff you need to do here), but maybe it helps to work with the XML files as long as the final solution to automatically pick up the right XSD within this new XML editor is released.

**Comment by boolivar** [comment]
@martinlippert I got this configuration working!

At first only Content Assist started working while Validator highlighted some errors.
I'm not sure, but it looks like it was complaining on elements in XSD that was imported from XSD I use, after IDE restart or whatever it started working.
I'll let you know if I encounter issues that matter.

Anyway thank you very much for your response and your time!

**Comment by martinlippert** [comment]
@boolivar Glad to hear you got it working. Did you use the catalog for it, as described above? Or did you figure out another way to make this work?

**Comment by boolivar** [comment]
@martinlippert I created a catalog based on your example
